File: Problems/0001_0099/0028_Implement_strStr/Project_CS/Implement_strStr.cs
using System;

public class Solution {
    public int StrStr_2(string haystack, string needle)
    {
        if ( needle == null || needle == "" )
        {
            return 0;
        }
    
        return(haystack.IndexOf(needle));
    }

    public int StrStr(string haystack, string needle)
    {
        if (needle.Length == 0) {
            return 0;
        }

        if (haystack.Length == 0) {
            return -1;
        }

        int i, j, n;
        
        for ( i = 0 ; i < haystack.Length; i++ ) {
            for ( n = i, j = 0; j < needle.Length; j++, n++ ) {
                if ( n >= haystack.Length ) {
                    return -1;
                }

                if ( haystack[n] != needle[j] ) {
                    break;
                }
            }
            
            // Console.WriteLine("i = " + i.ToString() + ", j = " + j.ToString() );

            if ( j == needle.Length ) {
                return i;
            }
        }
    
        return -1;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("\"","").Replace("[[","").Replace("]]","").Trim().Split(new string[] {"],["}, StringSplitOptions.None);
        string haystack = flds[0];
        string needle = flds[1];

        Console.WriteLine("haystack = " + haystack + "\n" + "needle = " + needle + "\n");

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = StrStr(haystack, needle);
        Console.WriteLine("result = " + result.ToString());

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0001_0099/0028_Implement_strStr/Project_Python3/Implement_strStr.py
import os
import sys
import time
from typing import List, Dict, Tuple

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        # 31ms - 38ms
        return haystack.find(needle)

    def strStr2(self, haystack: str, needle: str) -> int:
        # 32ms - 40ms
        for i, char in enumerate(haystack):
            if char == needle[0] and (i + len(needle) - 1) < len(haystack) and haystack[i:i + len(needle)] == needle:
                return i
        return -1

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip().split(",")
    haystack = flds[0]
    needle = flds[1]
    print("haystack = {0}, needle = {1}".format(haystack, needle))

    sl = Solution()
    time0 = time.time()

    result = sl.strStr(haystack, needle)

    time1 = time.time()

    print("result = {0}".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
